feat(useWorkerLoad): add new hook

fjcaetano opened the pull request:
fix #11

fjcaetano committed:
feat(useWorkerLoad): add new hook

fix #11

uqmessias commented on src/index.tsx:
```diff
@@ -179,3 +179,46 @@ export const useLazyRef = <T extends any>(
+ * @param initialValue The data's initial value. Optional when Data may be undefined
+ */
+export const useWorkerLoad = <Data extends any>(
+  worker: () => Promise<Data>,
```
What happens if the worker is a function that depends on other data? Shouldn't we have a `dependencies?: Dependencies` here?

uqmessias replied:
If the idea is to not allow the `worker` to be updated, shouldn't it be explained some how in the JSDoc?

fjcaetano replied:
The idea is that the `worker` function is only executed when the component gets mounted, so there's no need to update its dependencies. That's also why the callback is not exposed.

I added the `retry` thinking of a scenario with errors, but I don't know if it should exist when `error` is `undefined`. What do you think? The way that it is now, maybe `reload` or `rework` would be better names for it. If we decide to allow the user to call `retry` whenever they want, then we may need to add `worker` as a dependency of `useWorker`, for the case where the function is updated with new values.

danibonilha replied:
If the idea is to retry only on error, maybe creating a boolean `retryOnError` something like that and handling inside `useWorkerLoad` would avoid requiring dependencies.

fjcaetano replied:
But when would `retry` happen in that case? Immediately? I don't know if that's the best case. How would we prevent infinite retries? IMO, it'd be better if the user triggered a new retry

fjcaetano replied:
Idea: what if `error` was typed in the return as: `error?: { value: Error; retry: () => void }`. With the coalescing syntax, it'd be simple to access `error?.value` and `error?.retry()`

danibonilha replied:
🤔 I like this idea

uqmessias commented on src/index.tsx:
```diff
@@ -179,3 +179,46 @@ export const useLazyRef = <T extends any>(
+  }, []);
+
+  // start loading immediately
+  useDidMount(callback);
```
💯

fjcaetano committed:
chore(Dependencies): upgrade typescript to 3.7

fjcaetano committed:
chore(Dependencies): upgrade prettier to support TS 3.7

This is temporarily pointing prettier to a specific commit, since a new version hasn't been released yet

uqmessias commented on CHANGELOG.md:
```diff
@@ -1,3 +1,21 @@
+# [0.3.0-alpha1](https://github.com/fjcaetano/react-hook-utilities/compare/v0.2.1...v0.3.0-alpha1) (2019-11-11)
```
💯

uqmessias reviewed (approved):
Good stuff

fjcaetano commented:
**TODO**:
- [x] Docs for this hook are not being exported with distribution code

fjcaetano committed:
docs(README): add useWorkerLoad example
